maestro-cli/src/main/java/maestro/cli/session/MaestroSessionManager.kt

```diff
@@ -0,0 +1,265 @@
+/*
+ *
+ *  Copyright (c) 2022 mobile.dev inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *
+ */
+
+package maestro.cli.session
+
+import dadb.Dadb
+import dadb.adbserver.AdbServer
+import io.grpc.ManagedChannelBuilder
+import ios.idb.IdbIOSDevice
+import maestro.Maestro
+import maestro.cli.device.Device
+import maestro.cli.device.PickDeviceInteractor
+import maestro.cli.device.Platform
+import maestro.drivers.IOSDriver
+import java.util.UUID
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import kotlin.concurrent.thread
+
+object MaestroSessionManager {
+    private const val defaultHost = "localhost"
+    private const val idbPort = 10882
+
+    private val executor = Executors.newScheduledThreadPool(1)
+
+    fun <T> newSession(
+        host: String?,
+        port: Int?,
+        deviceId: String?,
+        block: (MaestroSession) -> T,
+    ): T {
+        val selectedDevice = selectDevice(host, port, deviceId)
+        val sessionId = UUID.randomUUID().toString()
+
+        val heartbeatFuture = executor.scheduleAtFixedRate(
+            {
+                SessionStore.heartbeat(sessionId, selectedDevice.platform)
+            },
+            0L,
+            10L,
+            TimeUnit.SECONDS
+        )
+
+        val session = SessionStore.withExclusiveLock {
+            createMaestro(
+                selectedDevice = selectedDevice,
+                connectToExistingSession = SessionStore.hasActiveSessions(sessionId, selectedDevice.platform),
+            )
+        }
+
+        Runtime.getRuntime().addShutdownHook(thread(start = false) {
+            SessionStore.withExclusiveLock {
+                SessionStore.delete(sessionId, selectedDevice.platform)
+
+                if (!SessionStore.hasActiveSessions(sessionId, selectedDevice.platform)) {
+                    session.close()
+                }
+            }
+        })
+
+        return try {
+            block(session)
+        } finally {
+            heartbeatFuture.cancel(true)
```

Comment:
Why not call `SessionStore.delete(sessionId, selectedDevice.platform)` here instead of secretly pruning sessions on a missing heartbeat?

Reply:
I actually just removed this `finally` section and moved everything into a shutdown hook. There is no strong guarantee that this block will be called if thread is interrupted.

Reply:
furthermore, pruning is needed in cases where even shutdown-hook have failed. We can't count on program to always execute this code successfully, so prunning is a last resort

Reply:
I get it, there is a separate thread that will make sure sessions are pruned -even if there is some unknown error in the application-